### PR TITLE
initial code to sync buildings placement during gameplay

### PR DESCRIPTION
**Pull request by sp00ktober**
This is just a start, there is still a lot missing. Whats working so far is:

- placement of buildings
- registering power generators and consumers to the power system
- connecting a miner to the correct veins
- flattening the terrain when placing a building and removing vegetation if needed.

It also includes a debug patch to free craft buildings and items.

### CODE REVIEW

File: NebulaPatcher/NebulaPatcher.csproj [hunk 1]
```diff
@@ -40,7 +40,8 @@
     <Reference Include="0Harmony">
       <HintPath>..\Libs\0Harmony.dll</HintPath>
     </Reference>
-    <Reference Include="Assembly-CSharp">
+    <Reference Include="Assembly-CSharp, Version=0.0.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
```
Comment (hubastard) [inline NebulaPatcher/NebulaPatcher.csproj:44]: Is this change is intentional ?

File: NebulaWorld/Factory/EntityManager.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            }
+            // place the entity
+            int ret = GameMain.mainPlayer.factory.AddEntityDataWithComponents(new EntityData
+            {
+                protoId = protoId,
+                pos = pos,
+                rot = rot
+            }, 0);
+
+            GameMain.mainPlayer.controller.actionBuild.NotifyBuilt(-0, ret);
```
Comment (hubastard) [inline NebulaWorld/Factory/EntityManager.cs:25]: I guess it should be `0` not `-0`

**Review by hubastard — APPROVED** [review]

File: NebulaPatcher/NebulaPatcher.csproj (hunk 1 above)
Comment (sp00ktober) [reply NebulaPatcher/NebulaPatcher.csproj:44]: not really 😅

File: NebulaModel/Packets/Factory/EntityPlaced.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using NebulaModel.DataStructures;
+using UnityEngine;
+
+namespace NebulaModel.Packets.Factory
+{
+    public class EntityPlaced
```
Comment (Therzok) [inline NebulaModel/Packets/Factory/EntityPlaced.cs:6]: This looks like another struct candidate

File: NebulaPatcher/Patches/Dynamic/Debugging.cs [hunk 4]
```diff
@@ -22,5 +23,29 @@ public static void Postfix(ref bool __result)
...
+    [HarmonyPatch(typeof(MechaForge), "AddTaskIterate")]
+    class patch4
+    {
+        public static bool Prefix(MechaForge __instance, ForgeTask __result, int recipeId, int count)
+        {
+            ForgeTask recipe = new ForgeTask(recipeId, count);
+            for(int i = 0; i < recipe.productIds.Length; i++)
+            {
+                GameMain.mainPlayer.package.AddItemStacked(recipe.productIds[i], count);
+            }
```
Comment (Therzok) [inline NebulaPatcher/Patches/Dynamic/Debugging.cs:45]: Shouldn't this add to the Mecha's storage, not the main player's?

Comment (Therzok) [reply NebulaPatcher/Patches/Dynamic/Debugging.cs:45]: Nevermind, this is debug code.

File: NebulaWorld/Factory/EntityManager.cs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            int veinsInAreaNonAlloc = collider.GetVeinsInAreaNonAlloc(center, 12f, tmp_ids);
+            VeinData[] veinPool = GameMain.mainPlayer.factory.veinPool;
+
+            int[] veinIDs = new int[veinsInAreaNonAlloc];
+            float veinOildClosest = 100f;
+            int oilVeinId = 0;
+
+            for (int i = 0; i < veinsInAreaNonAlloc; i++)
+            {
+                if(minerType == EMinerType.Vein)
```
Comment (Therzok) [inline NebulaWorld/Factory/EntityManager.cs:48]: Over here, maybe moving the miner type switch above the loop might be better.

Comment (Therzok) [reply NebulaWorld/Factory/EntityManager.cs:48]: I'll PR.

File: NebulaWorld/Factory/EntityManager.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                            GameMain.mainPlayer.factory.RefreshVeinMiningDisplay(minerPool[minerId].veins[i], entityId, 0);
+                        }
+                        minerPool[minerId].ArrageVeinArray();
+                        if (pcID != 0)
+                        {
+                            // this is hugely important to get power to the building!!!
+                            minerPool[minerId].pcId = pcID;
+                        }
+                        minerPool[minerId].GetMinimumVeinAmount(GameMain.mainPlayer.factory, GameMain.mainPlayer.factory.veinPool);
+                        // TODO: do some stuff with entitySignPool, is it important?
```
Comment (Therzok) [inline NebulaWorld/Factory/EntityManager.cs:169]: This is basically updating the image overlays on top of rendered objects. In example, logistics stations, recipes in assembler, etc.

**Pull request merged**